### ISSUE TITLE
Ancestor sequences reconstruction problem

### ISSUE THREAD

**Issue by purereason** [veg/hyphy#449]
Dear all,

I'm currently trying to reconstruct ancestral states on the phylogenetic tree using "-p" command line option (I'm using the mpi-version of the latest source code release).

I've modified AnalyzeNucDataFreq.bf batch file as shown below, but something goes wrong.

Here is my modified batch file:

```
SetDialogPrompt ("Please specify a nuleotide data or aminoacid file:");

DataSet ds = ReadDataFile (PROMPT_FOR_FILE);
DataSetFilter filteredData = CreateFilter (ds,1);

global estPiA;
global estPiC;
global estPiG;
global estPiT;
global sum;

estPiA:<1;
estPiC:<1;
estPiG:<1;
estPiT:=Abs(1-estPiA-estPiC-estPiG);
sum := estPiA+estPiC+estPiG+estPiT;

EMBED_FREQUENCY_DEPENDENCE = 1;

SelectTemplateModel(filteredData);

estPiA = .25;
estPiC = .25;
estPiG = .25;

vectorOfFrequencies = {{estPiA/sum},{estPiC/sum},{estPiG/sum},{estPiT/sum}};

fprintf (stdout,"\n",ds);

_DO_TREE_REBALANCE_ = 0;
ACCEPT_ROOTED_TREES = 1;

#include "queryTree.bf";

LikelihoodFunction lf = (filteredData,givenTree);

Optimize (res,lf);

fprintf (stdout, "\n",lf);

HarvestFrequencies (obsFreq,filteredData,1,1,1);

fprintf (stdout, "\nEstimated Frequencies:", vectorOfFrequencies, "\nObserved Frequencies", obsFreq);

fprintf (stdout,"\n",mbl,"\n");
```

I've (i) changed _DO_TREE_REBALANCE_ option's value as I wanted to preserve node positions (I'm aware that it is unrooted anyway, but it is more convenient for my purpose), (ii) deleted a string with "GetString (mbl, TrNModel,-1);" as it caused a fatal error and here http://www.hyphy.org/w/index.php/Template_Batch_Files it is written that this batch file "Does GetString() on TrNModel at end of file for no apparent reason".

I'm launching HyPhy as follows (GRM, 6 classes of rate heterogeneity):

```
mpirun -np 32 hyphympi -p /full_path/post_ancestors.bf /full_path/modified_batch.bf <<< $'/full_path/alignment.fasta\nGRM\n3\n8\n6\n/full_path/tree.nwc\ny\n7\n/full_path/reconstructed_ancestors.txt\nn' 1> stdout.txt 2> stderr.txt &
```

It works and outputs ancestral sequences. However, they are quite messy, with the part of them containing poly-A sequences instead of normal ones. When I'm using a simpler model (GRM without rate heterogeneity, which corresponds to "GRM\n2\n" option), everything seems fine, and I'm not observing any strange sequences.
In both cases, "messages.log" output files contain strings like this:

```
Internal error in _LikelihoodFunction::ConjugateGradientDescent. The function evaluated at current parameter values [-21351.81128983131] does not match the last recorded LF maximum [-21351.81128983132]
```

Including rate heterogeneity into a model causes lots of such error strings, while in simpler model I get only a single error.

I've tried other models with rate heterogeneity as well (TRN, HKY). I don't obtain any error messages in "messages.log" file, but still get poly-A sequences. For TRN model, some ancestral sequences consist of As and Ts, and some are polyA.

When I'm reducing the number of heterogeneity classes for GRM model (4 instead of 6), I don't obtain any errors, but obtain poly-A or [AT]n sequences.

Just in case that matters, I'm working with highly unbalanced influenza tree. I've tried to allow HYPHY to rebalance the tree (by removing "`ACCEPT_ROOTED_TREES = 1;`" string and changing _DO_TREE_REBALANCE_ variable back to 1). GRM model with 4 heterogeneity classes in this case still outputs many poly-A sequences. 

I've also noticed that HYPHY changes branch lengths making them much smaller, and many branches now have zero length. I failed to fix it; still, I'm not sure if it could cause my problems with ancestors.

I would be glad if you could help me to find out the reason for these observations. If possible, I'd like to retain the heterogeneity in my analysis.

Thank you in advance.

**Comment by spond** [comment]
Dear @purereason,

Could you confirm the version of HyPhy you are using? The issues with 'A' used to do with gappy sites, and should have been resolved in the last 12-24 months. Perhaps they remain with rate heterogeneity models -- I would need a complete test case from you (data and .hbf) to confirm. Can you please supply an example when you get poly-A sequences?

You can ignore the internal error warnings -- they are optimization routine diagnostics related to numerical instability. 

Typically you would set the  `AUTOMATICALLY_CONVERT_BRANCH_LENGTHS = 1; ` before you read any trees to instruct HyPhy to map branch lengths from the values you have in the tree to initial values for the optimizer. The branch lengths are optimized as parameters of the model, and reflect MLE values. They could be quite different from what you estimated previously, especially as you explore rate variation options.

Sergei

**Comment by purereason** [comment]
Dear @spond,

Many thanks for your response!

HyPhy version is called "/HYPHY 2.220160609beta(MPI) for Linux on x86_64\".

ZIP archive with example data and slightly modified AnalyzeNucDataFreq.bf (without ```"GetString (mbl, TrNModel,-1);"``` string) is attached. I think tree rerooting doesn't affect these poly-As so ```DO_TREE_REBALANCE``` variable isn't modified here.

When launching like this (GRM, general discrete, 4 classes of heterogeneity),
```mpirun -np 32 hyphympi -p /full_path/AnalyzeNucDataFreq.bf <<< $'/full_path/seq.fasta\nGRM\n3\n8\n4\n/full_path/tree_rooted.nwk\ny\n7\n/full_path/reconstructed_ancestors.txt\nn' 1> stdout.txt 2> stderr.txt &```
I obtain many poly-A ancestors. My alignment is gapless, btw. 

It turned out that ```[Gamma+Inv]``` option in GRM model (```GRM\n3\n5\n4``` option in command line) results in correct ancestors. It is computationally more expensive than discrete model, though.

Instead of AnalyzeNucDataFreq.bf, I tried AnalyzeNucProtData.bf and obtained the same results.

I've tested my dataset in GUI version as well; via "Standard Analyses" I obtain poly-As using AnalyzeNucDataFreq or AnalyzeNucProtData analyses under discrete GRM model, but good-looking ancestors under [Gamma+Inv] GRM model.
Via "New Analysis", by loading my alignment and tree into separate windows, then building and optimizing likelihood, I obtain normal ancestors under REV model, with 4 classes of rate heterogeneity and equilibrium frequencies taken from the partition. May be these observations can help. 

Thank you.

[hyphy_test.zip](https://github.com/veg/hyphy/files/372463/hyphy_test.zip)
